src/libvmmalloc.c

```diff
@@ -0,0 +1,409 @@
+ * will be redirected to libvmmalloc.
+ */
+void *(*__malloc_hook) (size_t size, const void *caller) =
+		(void *)malloc;
+void *(*__realloc_hook) (void *ptr, size_t size, const void *caller) =
+		(void *)realloc;
+void (*__free_hook) (void *ptr, const void *caller) =
+		(void *)free;
+void (*__memalign_hook) (size_t size, size_t alignment, const void *caller) =
+		(void *)memalign;
+#endif
+
+/*
+ * print_jemalloc_messages -- custom print function, for jemalloc
+ *
+ * Prints traces from jemalloc. All traces from jemalloc
+ * are considered as error messages.
+ */
+static void
+print_jemalloc_messages(void *ignore, const char *s)
+{
+	LOG_NONL(1, "%s", s);
+}
+
+/*
+ * libvmmalloc_create -- create a memory pool in a temp file
+ */
+static VMEM *
+libvmmalloc_create(const char *dir, size_t size)
+{
+	LOG(3, "dir \"%s\" size %zu", dir, size);
+
+	if (size < VMEM_MIN_POOL) {
+		LOG(1, "size %zu smaller than %zu", size, VMEM_MIN_POOL);
+		errno = EINVAL;
+		return NULL;
+	}
+
+	/* silently enforce multiple of page size */
+	size = roundup(size, Pagesize);
+
+	void *addr;
+	if ((addr = util_map_tmpfile(dir, size)) == NULL)
+		return NULL;
+
+	/* store opaque info at beginning of mapped area */
+	struct vmem *vmp = addr;
+	memset(&vmp->hdr, '\0', sizeof (vmp->hdr));
+	strncpy(vmp->hdr.signature, VMEM_HDR_SIG, POOL_HDR_SIG_LEN);
+	vmp->addr = addr;
+	vmp->size = size;
+	vmp->caller_mapped = 0;
+
+	/* Prepare pool for jemalloc */
+	if (je_vmem_pool_create((void *)((uintptr_t)addr + Header_size),
+			size - Header_size, 1) == NULL) {
+		LOG(1, "return NULL");
+		util_unmap(vmp->addr, vmp->size);
+		return NULL;
+	}
+
+	/*
+	 * If possible, turn off all permissions on the pool header page.
+	 *
+	 * The prototype PMFS doesn't allow this when large pages are in
+	 * use. It is not considered an error if this fails.
+	 */
+	util_range_none(addr, sizeof (struct pool_hdr));
+
+	LOG(3, "vmp %p", vmp);
+	return vmp;
+}
+
+/*
+ * libvmmalloc_init -- load-time initialization for libvmmalloc
+ *
+ * Called automatically by the run-time loader.
+ */
+__attribute__((constructor))
+static void
+libvmmalloc_init(void)
+{
+	char *size_str;
+	char *dir = NULL;
+	size_t size = 0;
+
+	out_init(VMMALLOC_LOG_PREFIX, VMMALLOC_LOG_LEVEL_VAR,
+			VMMALLOC_LOG_FILE_VAR);
+	out_set_vsnprintf_func(je_vmem_navsnprintf);
+	LOG(3, NULL);
+	util_init();
+
+	/* Set up jemalloc messages to a custom print function */
+	je_vmem_malloc_message = print_jemalloc_messages;
+
+	Header_size = roundup(sizeof (VMEM), Pagesize);
+
+	if ((dir = getenv(VMMALLOC_POOL_DIR_VAR)) == NULL) {
+		LOG(3, "%s not specified, using TMPDIR", VMMALLOC_POOL_DIR_VAR);
+		if ((dir = getenv("TMPDIR")) == NULL) {
+			LOG(3, "TMPDIR not specified, using \"/tmp\"");
+			dir = "/tmp";
+		}
+	}
+
+	if ((size_str = getenv(VMMALLOC_POOL_SIZE_VAR)) == NULL) {
+		LOG(3, "%s not specified, using default size (%zu)",
+			VMMALLOC_POOL_SIZE_VAR, VMEM_MIN_POOL);
+		size = VMEM_MIN_POOL;
```

Comment:
I'm going to change this.
If *VMMALLOC_POOL_SIZE* and/or *VMMALLOC_POOL_DIR* variables are not specified, it should be assumed as a fatal error. Or, if the library falls back to some default values, there must be at least a warning message printed on the console (regardless of the log level and build version). Otherwise, it leads to some annoying and unexpected failures, when the user does not export the environment variables properly or misspells their name.